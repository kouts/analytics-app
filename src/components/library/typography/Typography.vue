<script>
import { TypographyElements } from './typography.constants'

export default {
  name: 'Typography',
  props: {
    as: {
      type: String,
      default: 'p',
      validator(value) {
        return TypographyElements.includes(value)
      }
    },
    italic: {
      type: Boolean,
      default: false
    },
    bold: {
      type: Boolean,
      default: false
    },
    uppercase: {
      type: Boolean,
      default: false
    },
    underline: {
      type: Boolean,
      default: false
    },
    large: {
      type: Boolean,
      default: false
    },
    small: {
      type: Boolean,
      default: false
    }
  },
  computed: {
    isItalic() {
      return this.italic ? 'typography--italic' : ''
    },
    isBold() {
      return this.bold ? 'typography--bold' : ''
    },
    isUppercase() {
      return this.uppercase ? 'typography--uppercase' : ''
    },
    isUnderline() {
      return this.underline ? 'typography--underline' : ''
    },
    isLarge() {
      return this.large ? 'typography--large' : ''
    },
    isSmall() {
      return this.small ? 'typography--small' : ''
    }
  },
  render(createElement) {
    return createElement(
      this.as,
      {
        class: `typography typography--${this.as} ${this.isItalic} ${this.isBold} ${this.isUppercase} ${this.isUnderline} ${this.isLarge} ${this.isSmall}`
      },
      this.$slots.default
    )
  }
}
</script>
<style lang="scss" scoped>
.typography {
  font-family: 'Inter', system-ui, BlinkMacSystemFont, 'Segoe UI', Roboto, 'Helvetica Neue', Arial, sans-serif;
  color: inherit;
  padding: 0;
  margin: 0;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;

  &--h1,
  &--h2,
  &--h3,
  &--h4,
  &--h5,
  &--h6,
  &--bold {
    font-weight: 600 !important;
  }

  &--h1 {
    font-size: 32px;
    line-height: 48px;
  }
  &--h2 {
    font-size: 24px;
    line-height: 36px;
  }
  &--h3 {
    font-size: 20px;
    line-height: 30px;
  }
  &--h4 {
    font-size: 16px;
    line-height: 24px;
  }
  &--h5 {
    font-size: 14px;
    line-height: 20px;
  }
  &--h6 {
    font-size: 12px;
    line-height: 18px;
  }

  &--p,
  &--span {
    font-weight: 400;
    font-size: 14px;
    line-height: 20px;
  }

  &--large {
    font-size: 16px;
  }

  &--small {
    font-weight: 400;
    font-size: 12px;
    line-height: 20px;
  }

  &--italic {
    font-style: italic;
  }

  &--uppercase {
    text-transform: uppercase;
  }

  &--underline {
    text-decoration: underline;
  }
}
</style>
