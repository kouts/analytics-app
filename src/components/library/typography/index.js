export { default } from './Typography.vue'
