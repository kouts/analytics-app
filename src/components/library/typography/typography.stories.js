import Typography from './Typography.vue'

export default {
  title: 'Design System/Typography',
  component: Typography
}

export const Default = () => ({
  components: { Typography },
  template: `
  <div>
    <Typography>text-normal</Typography>
    <Typography italic>text-italic</Typography>
    <Typography bold>text-bold</Typography>
    <Typography uppercase>text-uppercase</Typography>
    <Typography bold uppercase>text-uppercase</Typography>
    <Typography underline>text-underline</Typography>

    <br/><br/>

    <Typography large>text-normal</Typography>
    <Typography large italic>text-italic</Typography>
    <Typography large bold>text-bold</Typography>
    <Typography large uppercase>text-uppercase</Typography>
    <Typography large bold uppercase>text-uppercase</Typography>
    <Typography large underline>text-underline</Typography>

    <br/><br/>

    <Typography as="small">text-normal</Typography><br/>
    <Typography as="small" italic>text-italic</Typography><br/>
    <Typography as="small" bold>text-bold</Typography><br/>
    <Typography as="small" uppercase>text-uppercase</Typography><br/>
    <Typography as="small" bold uppercase>text-uppercase</Typography><br/>
    <Typography as="small" underline>text-underline</Typography><br/>

    <br/><br/>

    <Typography as="h1">Heading 1</Typography>
    <Typography as="h2">Heading 2</Typography>
    <Typography as="h3">Heading 3</Typography>
    <Typography as="h4">Heading 4</Typography>
    <Typography as="h5">Heading 5</Typography>
    <Typography as="h6">Heading 6</Typography>
  </div>
`
})
