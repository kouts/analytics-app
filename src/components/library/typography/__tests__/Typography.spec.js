import Typography from '../Typography.vue'
import { TypographyElements } from '../typography.constants'
import { mount } from '@vue/test-utils'

describe('Typography', () => {
  it('should be defined', () => {
    expect(Typography).toBeDefined()
  })

  describe('snapshots', () => {
    it('should render default as p element', () => {
      const component = _mount({})

      expect(component).toMatchSnapshot()
    })

    it.each(TypographyElements)('should render %s element', (as) => {
      const component = _mount({ as })

      expect(component).toMatchSnapshot()
    })

    it.each(['italic', 'bold', 'uppercase', 'underline', 'large'])('should render with %s', (prop) => {
      const component = _mount({ [prop]: true })

      expect(component).toMatchSnapshot()
    })
  })

  /**
   * Simple wrapper to avoid duplicated code
   */
  const _mount = (props) => {
    return mount(Typography, {
      propsData: props,
      slots: {
        default: 'Slot content'
      }
    })
  }
})
