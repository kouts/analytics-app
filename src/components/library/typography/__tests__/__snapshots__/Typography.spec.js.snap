// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typography snapshots should render default as p element 1`] = `<p class="typography typography--p      ">Slot content</p>`;

exports[`Typography snapshots should render h1 element 1`] = `<h1 class="typography typography--h1      ">Slot content</h1>`;

exports[`Typography snapshots should render h2 element 1`] = `<h2 class="typography typography--h2      ">Slot content</h2>`;

exports[`Typography snapshots should render h3 element 1`] = `<h3 class="typography typography--h3      ">Slot content</h3>`;

exports[`Typography snapshots should render h4 element 1`] = `<h4 class="typography typography--h4      ">Slot content</h4>`;

exports[`Typography snapshots should render h5 element 1`] = `<h5 class="typography typography--h5      ">Slot content</h5>`;

exports[`Typography snapshots should render h6 element 1`] = `<h6 class="typography typography--h6      ">Slot content</h6>`;

exports[`Typography snapshots should render p element 1`] = `<p class="typography typography--p      ">Slot content</p>`;

exports[`Typography snapshots should render small element 1`] = `<small class="typography typography--small      ">Slot content</small>`;

exports[`Typography snapshots should render span element 1`] = `<span class="typography typography--span      ">Slot content</span>`;

exports[`Typography snapshots should render with bold 1`] = `<p class="typography typography--p  typography--bold    ">Slot content</p>`;

exports[`Typography snapshots should render with italic 1`] = `<p class="typography typography--p typography--italic     ">Slot content</p>`;

exports[`Typography snapshots should render with large 1`] = `<p class="typography typography--p     typography--large ">Slot content</p>`;

exports[`Typography snapshots should render with underline 1`] = `<p class="typography typography--p    typography--underline  ">Slot content</p>`;

exports[`Typography snapshots should render with uppercase 1`] = `<p class="typography typography--p   typography--uppercase   ">Slot content</p>`;
